packages/zoe/src/contractSupport/ratio.js

```diff
@@ -0,0 +1,171 @@
+import './types';
+import { assert, details as X } from '@agoric/assert';
+import Nat from '@agoric/nat';
+import { natSafeMath } from './safeMath';
+
+const { multiply, floorDivide, subtract } = natSafeMath;
+
+// make a Ratio, which represents a fraction. It is a pass-by-copy record.
+//
+// The natural syntax for the most common operations we want to support
+// are Amount * Ratio and Amount / Ratio. Since the operations want to adhere to
+// the ratio rather than the amount, we settled on a calling convention of
+// multiplyBy(Amount, Ratio) and divideBy(Amount, Ratio).
+//
+// The most common kind of Ratio can be applied to Amounts of a particular
+// brand, and produces results of the same brand. This represents a multiplier
+// that is only applicable to that brand. The less common kind of Ratio can be
+// applied to one particular brand of amounts, and produces results of another
+// particular brand. This represents some kind of exchange rate. The
+// brand-checking helps us ensure that normal Ratios aren't applied to amounts
+// of the wrong brand, and that exchange rates are only used in the appropriate
+// direction.
+
+const BASIS_POINTS = 10000;
+const PERCENT = 100;
+
+const ratioPropertyNames = [
+  'numerator',
+  'denominator',
+  'numeratorBrand',
+  'denominatorBrand',
+];
+
+function assertIsRatio(ratio) {
+  const propertyNames = Object.getOwnPropertyNames(ratio);
+  assert(propertyNames.length === 4, X`Ratio must be a record with 4 fields`);
+  for (let i = 0; i < 4; i += 1) {
+    assert(
+      ratioPropertyNames[i] === propertyNames[i],
+      X`parameter must be a Ratio record`,
+    );
+  }
+  Nat(ratio.numerator);
+  Nat(ratio.denominator);
+}
+
+export function makeRatio(
+  numerator,
+  numeratorBrand,
+  denominator = PERCENT,
+  denominatorBrand = numeratorBrand,
+) {
+  Nat(numerator);
+  Nat(denominator);
+  assert(denominator > 0, X`No infinite ratios!`);
+
+  return harden({
+    numerator,
+    denominator,
+    numeratorBrand,
+    denominatorBrand,
+  });
+}
+
+export function makeRatioFromAmounts(
+  numeratorAmount,
+  denominatorAmount,
+  base = BASIS_POINTS,
+) {
+  assert(denominatorAmount.value > 0, X`No infinite ratios!`);
+
+  const numerator = floorDivide(
+    multiply(base, numeratorAmount.value),
+    denominatorAmount.value,
+  );
```

Comment:
Given the parallelism of the names, I would expect
```js
makeAmountFromRatio(moola(5n), bucks(3n))
```
to produce
```js
{
  numerator: 5n,
  denominator: 3n,
  numeratorBrand: moolaBrand,
  denominatorBrand: buckBrand,
}
```
exactly as if I had called
```js
makeRatio(5n, 3n, moolaBrand, bucksBrand)
```
Instead, IIUC, I'd get
```js
{
  numerator: 16666n,
  denominator: 10000n,
  numeratorBrand: moolaBrand,
  denominatorBrand: buckBrand,
}
```

Reply:
Thanks. Base is a vestige of thinking about this in terms of percents. As you say, we don't need it at all.

Reply:
Also we are changing to just having 2 amounts now?